### ISSUE TITLE
reauthentication gets stuck in a mutex deadlock

### ISSUE THREAD

**Issue by majewsky** [gophercloud/gophercloud#1459]
Since merging #1226, reauthentication gets stuck in a mutex deadlock. To reproduce:

```go
package main

import (
	"fmt"

	"github.com/gophercloud/gophercloud"
	"github.com/gophercloud/gophercloud/openstack"
	"github.com/gophercloud/gophercloud/openstack/identity/v3/projects"
	"github.com/gophercloud/utils/openstack/clientconfig"
)

func failIf(err error) {
	if err != nil {
		panic(err.Error())
	}
}

func main() {
	authOpts, err := clientconfig.AuthOptions(nil)
	failIf(err)
	authOpts.AllowReauth = true
	provider, err := openstack.AuthenticatedClient(*authOpts)
	failIf(err)

	//simulate expired token by setting a bogus token value
	provider.SetToken("this is not a valid token")

	client, err := openstack.NewIdentityV3(provider, gophercloud.EndpointOpts{})
	failIf(err)

	//do a random request -> this gets a 401 response, does reauth, and gets stuck during reauth
	pages, err := projects.List(client, nil).AllPages()
	failIf(err)
	projects, err := projects.ExtractProjects(pages)
	failIf(err)
	fmt.Printf("There are %d projects.\n", len(projects))
}
```

What happens, as far as I can see:

1. ProviderClient.Reauthenticate locks `client.mut`.
2. ProviderClient.Reauthenticate calls its ReauthFunc.
3. Upon receiving the authentication result, the ReauthFunc calls ProviderClient.SetTokenAndAuthResult.
4. ProviderClient.SetTokenAndAuthResult attempts to lock `client.mut` again, thus getting stuck in a deadlock.

I'm not sure though why this problem only appeared after #1226. It didn't substantially change the way that `client.mut` was used. However, I suspect that there is some interaction with #1408.

@zimnx @jrperritt Any ideas what's going on here? In particular: What are the semantics of `client.reauthmut` and `client.mut`, respectively? What are these mutexes supposed to guard exactly? Do we have some documentation on this somewhere?

@jtopjian Since this is a pretty severe issue (and it hit me very hard in head in production), I'm totally fine with you reverting #1226 if this cannot be resolved quickly.

**Comment by majewsky** [comment]
As a general note, I think the reauth code has gotten way too complex for its own good. Maybe we should do away with the mutexes entirely, and use channels instead? (i.e. the ProviderClient creates a goroutine for the reauth at creation time, and when it (or its ServiceClients) wants to reauth, they send a message to that goroutine to do the reauth for them)

**Comment by jtopjian** [comment]
> As a general note, I think the reauth code has gotten way too complex for its own good.

I agree. I'm tempted to revert all reauth stuff to how it was 1-2 years ago when the user had to supply the entire reauth logic. There has been a cat and mouse game going on for too long to support everyone's use-case.

> Any ideas what's going on here? In particular: What are the semantics of `client.reauthmut` and `client.mut`, respectively?

Each have a legitimate role. I need to walk through the code to remind myself, but I do recall that there were reasons for having them both.

I will try to look into this ASAP. If I don't get to it, I'll revert #1226 temporarily.

**Comment by jtopjian** [comment]
This _should_ be fixed in #1464, but let me know if you're still having problems.
